### PR TITLE
Updates to ios Storage section

### PR DESCRIPTION
**Pull request by cbommas**

### CODE REVIEW

File: ios/storage.md [hunk 1]
```diff
@@ -161,317 +161,620 @@ func downloadData() {
...
-    file: S3ObjectInput!
-    version: Int!
-}
+This option is set to false by default.
 
-input DeletePostInput {
-    id: ID!
-}
+```swift
+//Setup credentials
```
Comment (undefobj) [inline ios/storage.md]: Can we put the `credentialsProvider` and `configuration` object in a separate section called "Manual Configuration" and point out that they should use the `AWSMobileClient` to manage all auth state, and link to that section in the docs directly. Then say something like "However, if you would like to manually provide your Cognito provider and Region you can do so with the following commands..."

We really want to start pointing people to use the new auth tooling with the AWSMobileClient now that it has all these features to ease on-boarding and other customer issues.

File: ios/storage.md [hunk 2]
```diff
@@ -161,317 +161,620 @@ func downloadData() {
...
+    for task in multiPartUploadTasks! {
+        task.setCompletionHandler(completionHandler!)
+        task.setProgressBlock(progressBlock!)
+
+ }
+```
+
+### Transfer with Object Metadata
+
+The `AWSS3TransferUtilityUploadExpression` and `AWSS3TransferUtilityMultiPartUploadExpression` classes contain the method `setValue:forRequestHeader` where you can pass in metadata to Amazon S3. This example demonstrates passing in the Server-side Encryption Algorithm as a request header in uploading data to S3 using MultiPart.
```
Comment (undefobj) [inline ios/storage.md]: We should link to official S3 docs here: https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html

File: ios/storage.md [hunk 3]
```diff
@@ -161,317 +161,620 @@ func downloadData() {
...
+                }
+
+                return nil;
+            }
 ```
 
-**Response Mapping Template**
+### Transfer with Access Control List
+
+To upload a file and specify permissions for it, you can use predefined grants, also known as canned ACLs. The following code shows you how to setup a file with publicRead access using the AWSS3 client.
```
Comment (undefobj) [inline ios/storage.md]: This is a pretty complex feature and arguably I'm not sure how good this is in a Mobile app as our "front page" in documentation, but let's discuss. If we do keep this we should link to the S3 docs: https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html

File: ios/storage.md [hunk 4]
```diff
@@ -161,317 +161,620 @@ func downloadData() {
...
+})
 ```
-$util.toJson($util.dynamodb.fromS3ObjectJson($context.source.file))
+
+### Downloading to a File
+
+```swift
+let fileURL = // The file URL of the download destination.
+
+let transferUtility = AWSS3TransferUtility.default()
```
Comment (undefobj) [inline ios/storage.md]: Isn't this redundant with the other Download examples above?

File: ios/storage.md [hunk 5]
```diff
@@ -161,317 +161,620 @@ func downloadData() {
...
+
+            if let _ = task.result {
+                // Do something with downloadTask.
+            }
+
+            return nil;
         }
+```
+
+### Limitations
```
Comment (undefobj) [inline ios/storage.md]: Let's chat about this section and the next one Monday. I'm not sure I would call this a "Limitations" section as much as a "How it Works" that perhaps we should move higher in this doc. Also I don't know if "Building a Pre-Signed URL" is really needed, nor if we need the SSE section if we already have it in the earlier section on Metadata.

File: ios/storage.md [hunk 6]
```diff
@@ -12,9 +12,9 @@ The CLI configures three different access levels on the storage bucket; public,
 
 - Files with public access level can be accessed by all users who are using your app. In S3, they are stored under the ``public/`` path in your S3 bucket.
 
-- Files with protected access level are readable by all users but writable only by the creating user. In S3, they are stored under ``protected/{user_identity_id}/`` where the user_identity_id corresponds to a unique Amazon Cognito Identity ID for that user.
+- Files with protected access level are readable by all users but writable only by the creating user. In S3, they are stored under ``protected/{user_identity_id}/`` where the ``user_identity_id`` corresponds to a unique Amazon Cognito Identity ID for that user.
 
-- Files with private access level are only accessible for specific authenticated users only. In S3, they are stored under ``private/{user_identity_id}/`` where the user_identity_id corresponds to a unique Amazon Cognito Identity ID for that user.
+- Files with private access level are only accessible for specific authenticated users only. In S3, they are stored under ``private/{user_identity_id}/`` where the ``user_identity_id`` corresponds to a unique Amazon Cognito Identity ID for that user.
```
[the remark this answers is not in the document]
Comment (cbommas) [reply ios/storage.md]: Updated the word-smithed rev of the text. Chose not to include the code snippet, as it introduces concepts such as AWSMobileClient and transferUtility out of context - lets chat on where best to put this sample in

File: ios/storage.md [hunk 7]
```diff
@@ -64,9 +64,9 @@ The CLI configures three different access levels on the storage bucket; public,
 
 ### Connect to Your Backend
 
-Use the following steps to connect add file storage backend services to your app.
+Use the following steps to add file storage backend services to your app.
 
-1. Add the following to `Podfile` that you configure to install the AWS Mobile SDK:
+1. Add the following to the `Podfile` to install the AWS Mobile SDK:
```
[the remark this answers is not in the document]
Comment (cbommas) [reply ios/storage.md]: Done

File: ios/storage.md (hunk 2 above)
Comment (cbommas) [reply ios/storage.md]: Done

File: ios/storage.md (hunk 3 above)
Comment (cbommas) [reply ios/storage.md]: Link added. Lets chat on whether to keep or remove.

File: ios/storage.md (hunk 5 above)
Comment (cbommas) [reply ios/storage.md]: Agreed - In fact I think we should remove the sections titled "Downloading to a file", "Uploading binary data to a file", "Downloading binary data to a file"

File: ios/storage.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (cbommas) [reply ios/storage.md]: Added a link to the Authentication section.

File: ios/storage.md (hunk 3 above)
Comment (cbommas) [reply ios/storage.md]: Removed as discussed.

File: ios/storage.md (hunk 4 above)
Comment (cbommas) [reply ios/storage.md]: Removed.

File: ios/storage.md (hunk 5 above)
Comment (cbommas) [reply ios/storage.md]: Changed the "Limitations" section to "How it works". Removed  SSE

File: ios/storage.md (hunk 1 above)
Comment (cbommas) [reply ios/storage.md]: Added a link to the authentication module.

**Pull request merged**